README
Redis::CappedCollection package version 1.09
============================================

Redis::CappedCollection module provides fixed sized collections that have
an auto-FIFO age-out feature. Collection size is determined by Redis server
'maxmemory' setting.

The Redis::CappedCollection package was written, tested, and found working on
recent Linux distributions.

INSTALLATION

To install this package type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

Perl version 5.010 or better is needed in order to install and use this package.
The Redis::CappedCollection module depends on other packages that are distributed
separately from Perl. We recommend that you have the following packages installed
before you install Redis::CappedCollection :

   Const::Fast
   Digest::SHA1
   Mouse
   Params::Util
   Redis
   Try::Tiny

The Redis::CappedCollection module has the following optional dependencies:

   Data::UUID
   JSON::XS
   Net::EmptyPort
   Sys::SigAction
   Test::Exception
   Test::NoWarnings
   Test::RedisServer

If the optional modules are missing, some "prereq" tests are skipped.

The installation of the missing dependencies can be accomplished either
through your OS package manager or through CPAN (or downloading the source
for all dependencies and compiling them manually).

TESTING

A running Redis server is required before launching tests.

We strongly recommend setting 'maxmemory' option in the redis.conf file.

TODO

Implement tests for
- memory errors handling (working with internal ROLLBACK commands).
- working with maxmemory = 0 in the redis.conf file.

SIMPLE BENCHMARK

Use benchmark_capped.pl to run a simple benchmark.

DOCUMENTATION

The Redis::CappedCollection module has a man page.
A simple description of the module is included below in the SEE ALSO section.

SEE ALSO

The basic operation of the Redis::CappedCollection package module:

   Redis::CappedCollection - Object interface to create a collection, addition
                             of data and data manipulation.
   Redis::CappedCollection::Util - String manipulation utilities.

COPYRIGHT AND LICENCE

Copyright (C) 2012-2016 by TrackingSoft LLC.

This package is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See perlartistic at
http://dev.perl.org/licenses/artistic.html .

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.


Changes
Revision history for Redis::CappedCollection module

    CHANGES:
    - max_list_items added.
    - Needs Redis server version 2.8 or higher.
    - Attempt to protect against used_memory > maxmemory.
    FIXES:
    - Obey 'connection_timeout', 'operation_timeout' attributes of redis connection hash.
    - Tests refined.

1.09  Tue Jun  14 2016
    CHANGES:
    - Now we use 'croak' instead 'confess'.
    - 'min_cleanup_bytes' and 'min_cleanup_items' renamed to
      'cleanup_bytes' and 'cleanup_items'.
    - $DEFAULT_MIN_CLEANUP_BYTES, $DEFAULT_MIN_CLEANUP_ITEMS renamed to
      $DEFAULT_CLEANUP_BYTES, $DEFAULT_CLEANUP_ITEMS .
    BUGFIXES:
    - Data::Dumper 'Sparseseen' method not used now.

1.08  Fri May  06 2016
    FIXES:
    - Tests corrected.

1.07  Tue May  03 2016
    FIXES:
    - Makefile.PL corrected.
    - Tests corrected.

1.06  Thu Apr  28 2016
    CHANGES:
    - Ability to force re-connection with Redis on error ('reconnect_on_error' attribute).
    - $DEFAULT_CONNECTION_TIMEOUT, $DEFAULT_OPERATION_TIMEOUT,
      $DEFAULT_ADVANCE_CLEANUP_NUM constants added.
    - connection_timeout, operation_timeout attributes added.
    - Revised error codes.
    - Better handling of errors when running out of memory.
    - Arguments control improved.
    - Added the service module Redis::CappedCollection::Util.
    - Avoid "cache stampede" effect during advanced memory cleanup.
    - Attributes 'advance_cleanup_bytes' and 'advance_cleanup_num' renamed to
      'min_cleanup_bytes' and 'min_cleanup_items'.
    - $DEFAULT_ADVANCED_CLEANUP_BYTES, $DEFAULT_ADVANCE_CLEANUP_NUM renamed to
      $DEFAULT_MIN_CLEANUP_BYTES, $DEFAULT_MIN_CLEANUP_ITEMS .
    FIXES:
    - Argument 'redis' may have Redis::CappedCollection type in function
      collection_info, collection_exists, drop_collection, redis_config_ok,
      resize calls.
    - Tests corrected.

1.05  Tue Jan  14 2016
    FIXES:
    - Error on update/upsert without $data_time in some cases (e.g. after pop_oldest).

1.04  Tue Jan  12 2016
    CHANGES:
    - Added 'upsert' method.

1.03  Thu Jan  07 2016
    CHANGES:
    - Works with a Redis server 'maxmemory-policy' setting 'noeviction' only.
    - Added 'redis_config_ok' method.
    - $E_REDIS_DID_NOT_RETURN_DATA, $E_UNKNOWN_ERROR errors added.
    - Updated version of the data structure.
    FIXES:
    - Incorrect analysis an error related to not enough memory.

1.02  Thu Sep  03 2015
    CHANGES:
    - Redis::CappedCollection does not close connection to redis server (including '->quit' method)
      if it was provided by user.
    - More convenient error codes naming.
    FIXES:
    - Fixes of test plans.

1.01  Tue Jul  07 2015
    CHANGES:
    - Updated behaviour of 'older_allowed' attribute that controls adding an element that's
      older than the last element removed from collection.
    - Added 'clear_collection' method.
    FIXES:
    - Small correction in destruction procedure.
    - Fixes of documentation and MANIFEST.

1.00  Mon Jun  15 2015

    !!! This is a major upgrade that changes API. It's strongly advised to study
    revised documentation and update your code where necessary !!!

    !!! Data structures created on the server by previous versions of the module
    are incompatible with this release. Existing collections must be recreated
    after updating to this release. !!!

    CHANGES:
    - Redis::CappedCollection->new() is deprecated and will be removed in future.
      Please use either create() or open() instead.
    - In the parameters of the methods 'redis' may be either a Redis object
      or a hash reference of parameters to create a Redis object.
    - Changed the order of the arguments of the 'insert' method.
    - The list ID and data ID arguments are mandatory for 'insert' method.
    - Added 'data time' argument for 'update' method.
    - Method 'info' renamed to 'list_info'.
    - Method 'exists' renamed to 'list_exists'.
    - Method 'drop' renamed to 'drop_list'.
    - Constructor parameter 'big_data_threshold' was removed.
    - 'last_removed_time' information attribute was removed.
    - Added constructor argument 'memory_reserve'.
    - Added 'oldest_time' method.
    - Use Const::Fast for values exported by the module.
    - Attributes 'advance_cleanup_bytes' and 'advance_cleanup_num' are stored
      on the Redis server and can be changed only by 'resize' method.
    - $MIN_MEMORY_RESERVE, $MAX_MEMORY_RESERVE constants added.
    - $ENONEXISTENTDATAID, $EINCOMPDATAVERSION errors added.
    - Redis server version 2.8 or higher is required.
    - We strongly recommend setting 'maxmemory' option in the redis.conf file.
    - Added utilities to dump and restore the collection: dump_collection,
      restore_collection .
    FIXES:
    - 'open' method fixed.
    - Tests fixed.

0.17  Thu Apr  23 2015
    CHANGES:
    - Added 'create', 'open' constructors.
    - Added 'collection_exists', 'resize' methods.
    - 'open', 'collection_info', 'collection_exists', 'resize', 'drop_collection'
      may be used as functions.
    - Redis::CappedCollection->new() is deprecated and will be removed in future.
      Please use either create() or open() instead.
    - The 'quit' method does not close connection to Redis during
      global destruction phase (Perl version 5.14 and above).

0.16  Tue May  27 2014
    CHANGES:
    - Added argument 'check_maxmemory' to the Redis::CappedCollection->new .

0.13 - 0.15  Wed Apr  02 2014
    CHANGES:
    - Small test improvement.

0.12  Fri Mar  14 2014
    CHANGES:
    - Change the port used for testing.

0.11  Fri Nov  15 2013
    CHANGES:
    - Adjust Redis server connection parameters.
      (Starting with version 1.963, Redis module doesn't support encoding)

0.10  Fri Oct  25 2013
    FIXES:
    - Fix runtime loading of Test::*

0.09  Fri Sep  27 2013
    FIXES:
    - Tests fine-tuned to the correct execution of 32-bit systems.

0.08  Wed Sep  18 2013
    CHANGES:

t/tlib/Redis/CappedCollection/Test/Utils.pm
package Redis::CappedCollection::Test::Utils;

use 5.010;
use strict;
use warnings;

use Exporter qw(
    import
);
our @EXPORT_OK  = qw(
    clear_coll_data
    get_redis
    verify_redis
);

use Net::EmptyPort;
use Test::More;
use Test::RedisServer;
use Try::Tiny;

use Redis::CappedCollection qw(
    $DEFAULT_SERVER
    $DEFAULT_PORT
    $NAMESPACE
);

sub get_redis
{
    my @args = @_;

    my ( $redis, $error );
    for ( 1..3 )
    {
        try
        {
            $redis = Test::RedisServer->new( @args );
        }
        catch
        {
            $error = $_;
        };
        last unless $error;
        sleep 1;
    }

    return wantarray ? ( $redis, $error ) : $redis;
}

sub verify_redis
{
    my $redis;
    my $real_redis;
    my $skip_msg;
    my $port = Net::EmptyPort::empty_port( $DEFAULT_PORT );

    $redis = get_redis(
        conf    => {
            port                => $port,
            'maxmemory-policy'  => 'noeviction',
        },
        timeout => 3,
    );
    if ( $redis )
    {
        eval { $real_redis = Redis->new( server => $DEFAULT_SERVER.":".$port ) };
        $skip_msg = "Redis server is unavailable" unless ( !$@ && $real_redis && $real_redis->ping );
        $skip_msg = "Need a Redis server version 2.8 or higher" if ( !$skip_msg && !eval { return $real_redis->eval( 'return 1', 0 ) } );
        unless ( $skip_msg ) {
            $real_redis->script_flush;
            $real_redis->quit;
        }
    }
    else
    {
        $skip_msg = "Unable to create test Redis server";
    }

    return $redis, $skip_msg, $port;
}

sub clear_coll_data {
    my ( $coll ) = @_;

    eval { $coll->_call_redis( 'DEL', $NAMESPACE.':Q:'.$coll->name ); };
    eval { $coll->_call_redis( 'DEL', $NAMESPACE.':S:'.$coll->name ); };
    eval { $coll->_call_redis( 'DEL', $coll->_call_redis( 'KEYS', $NAMESPACE.':I:'.$coll->name.":*" ) ); };
    eval { $coll->_call_redis( 'DEL', $coll->_call_redis( 'KEYS', $NAMESPACE.':D:'.$coll->name.":*" ) ); };
    eval { $coll->_call_redis( 'DEL', $coll->_call_redis( 'KEYS', $NAMESPACE.':T:'.$coll->name.":*" ) ); };
}

1;


example.pl
#!/usr/bin/perl -w

# NAME: Redis CappedCollection demonstration

use 5.010;
use strict;
use warnings;

#-- Common ---------------------------------------------------------------------
use Redis::CappedCollection qw(
    $DEFAULT_SERVER
    $DEFAULT_PORT

    $E_NO_ERROR
    $E_MISMATCH_ARG
    $E_DATA_TOO_LARGE
    $E_NETWORK
    $E_MAXMEMORY_LIMIT
    $E_COLLECTION_DELETED
    $E_REDIS
    $E_DATA_ID_EXISTS
    $E_OLDER_THAN_ALLOWED
    );

my $server = $DEFAULT_SERVER.":".$DEFAULT_PORT;   # the Redis Server

sub exception {
    my $coll    = shift;
    my $err     = shift;

    die $err unless $coll;
    if ( $coll->last_errorcode == $E_NO_ERROR )
    {
        # For example, to ignore
        return unless $err;
    }
    elsif ( $coll->last_errorcode == $E_MISMATCH_ARG )
    {
        # Necessary to correct the code
    }
    elsif ( $coll->last_errorcode == $E_DATA_TOO_LARGE )
    {
        # You must use the control data length
    }
    elsif ( $coll->last_errorcode == $E_NETWORK )
    {
        # For example, sleep
        #sleep 60;
        # and return code to repeat the operation
        #return 'to repeat';
    }
    elsif ( $coll->last_errorcode == $E_MAXMEMORY_LIMIT )
    {
        # For example, return code to restart the server
        #return 'to restart the redis server';
    }
    elsif ( $coll->last_errorcode == $E_COLLECTION_DELETED )
    {
        # For example, return code to ignore
        #return "to ignore $err";
    }
    elsif ( $coll->last_errorcode == $E_REDIS )
    {
        # Independently analyze the $err
    }
    elsif ( $coll->last_errorcode == $E_DATA_ID_EXISTS )
    {
        # For example, return code to reinsert the data
        #return "to reinsert with new data ID";
    }
    elsif ( $coll->last_errorcode == $E_OLDER_THAN_ALLOWED )
    {
        # Independently analyze the situation
    }
    else
    {
        # Unknown error code
    }
    die $err if $err;
}

my ( $id, $coll, @data );

eval {
    $coll = Redis::CappedCollection->create(
        redis   => $server,
        name    => 'Some name', # Create a collection with the specified name
        );
};
exception( $coll, $@ ) if $@;
print "'", $coll->name, "' collection created.\n";

#-- Producer -------------------------------------------------------------------
#-- New data

eval {
    $id = $coll->insert(
        'Some_unique_id',
        123,                # data id
        'Some data stuff',
        );
    print "Added data in a list with '", $id, "' id\n";

    if ( $coll->update( $id, 0, 'Some new data stuff' ) )
    {
        print "Data updated successfully\n";
    }
    else
    {
        print "The data is not updated\n";
    }
};
exception( $coll, $@ ) if $@;

#-- Consumer -------------------------------------------------------------------
#-- Fetching the data

eval {
    @data = $coll->receive( $id );
    print "List '$id' has '$_'\n" foreach @data;

# or
    while ( my ( $id, $data ) = $coll->pop_oldest )
    {
        print "List '$id' had '$data'\n";
    }
};
exception( $coll, $@ ) if $@;

#-- Utility --------------------------------------------------------------------
#-- Getting statistics

my ( $lists, $items );
eval {
    my $info = $coll->collection_info;
    print 'An existing collection uses: ',
        $info->{items}, ' items are placed in ',
        $info->{lists}, ' lists', "\n";

    print "The collection has '$id' list\n" if $coll->list_exists( 'Some_id' );

    print "Collection '", $coll->name, "' has '$_' list\n" foreach $coll->lists;
};
exception( $coll, $@ ) if $@;

#-- Closes and cleans up -------------------------------------------------------

eval {
    $coll->quit;

# Before use, make sure that the collection is not being used by other customers
#    $coll->drop_collection;
};
exception( $coll, $@ ) if $@;

exit;


coverage.sh
cover -delete

#find t -name \*.t -print -exec perl -MDevel::Cover {} \;

perl -MDevel::Cover example.pl

perl -MDevel::Cover t/00_compile.t
perl -MDevel::Cover xt/author/92_vars.t
perl -MDevel::Cover xt/author/93_versions.t
perl -MDevel::Cover xt/author/94_fixme.t
perl -MDevel::Cover xt/author/95_critic.t
perl -MDevel::Cover xt/release/96_kwalitee.t
perl -MDevel::Cover xt/release/97_distribution.t
perl -MDevel::Cover xt/release/98_pod.t
perl -MDevel::Cover xt/release/99_pod_coverage.t

perl -MDevel::Cover t/00_CappedCollection/00_compile.t
perl -MDevel::Cover t/00_CappedCollection/01_new.t
perl -MDevel::Cover t/00_CappedCollection/02_insert.t
perl -MDevel::Cover t/00_CappedCollection/03_update.t
perl -MDevel::Cover t/00_CappedCollection/04_receive.t
perl -MDevel::Cover t/00_CappedCollection/05_collection_info.t
perl -MDevel::Cover t/00_CappedCollection/06_pop_oldest.t
perl -MDevel::Cover t/00_CappedCollection/07_exists.t
perl -MDevel::Cover t/00_CappedCollection/08_lists.t
perl -MDevel::Cover t/00_CappedCollection/09_drop_collection.t
perl -MDevel::Cover t/00_CappedCollection/10_quit.t
perl -MDevel::Cover t/00_CappedCollection/11_max_datasize.t
perl -MDevel::Cover t/00_CappedCollection/12_last_errorcode.t
perl -MDevel::Cover t/00_CappedCollection/13_name.t
perl -MDevel::Cover t/00_CappedCollection/15_cleanup_bytes.t
perl -MDevel::Cover t/00_CappedCollection/17_cleanup_items.t
perl -MDevel::Cover t/00_CappedCollection/18_info.t
perl -MDevel::Cover t/00_CappedCollection/19_drop.t
perl -MDevel::Cover t/00_CappedCollection/20_cleaning.t
perl -MDevel::Cover t/00_CappedCollection/21_rollback.t
perl -MDevel::Cover t/00_CappedCollection/22_alarm.t
perl -MDevel::Cover t/00_CappedCollection/23_cleaning_bench.t
perl -MDevel::Cover t/00_CappedCollection/24_timeout.t
perl -MDevel::Cover t/00_CappedCollection/25_cleaning_correctness.t
perl -MDevel::Cover t/00_CappedCollection/26_script_variable_protection.t
perl -MDevel::Cover xt/release/00_CappedCollection/98_pod.t
perl -MDevel::Cover xt/release/00_CappedCollection/99_pod_coverage.t

cover


benchmark_capped.pl
#!/usr/bin/perl -w

use 5.010;
use strict;
use warnings;

use lib qw(
    lib
    t/tlib
);

# NAME: Redis::CappedCollection benchmark

use bytes;
use Carp;
use Time::HiRes     qw( gettimeofday );
use Redis '1.976';
use List::Util      qw( min sum );
use Getopt::Long    qw( GetOptions );
use Data::Dumper;

use Redis::CappedCollection qw(
    $DEFAULT_SERVER
    $DEFAULT_PORT
);

# ENVIRONMENT ------------------------------------------------------------------

#-- declarations ---------------------------------------------------------------

use constant {
    PORTION_TIME            => 10,
    INFO_TIME               => 60,
    VISITOR_ID_LEN          => 20,
    LOST_PRODUCTIVITY       => 3,               # allowable percentage of lost productivity
    };

my $host                    = $DEFAULT_SERVER;
my $port                    = $DEFAULT_PORT;
my $empty_port;
my $server;
my $redis;
my $max_size                = 512 * 1024 * 1024;    # 512MB
my $min_cleanup_bytes       = 0;
my $min_cleanup_items       = 0;
my $max_lists               = 2_000_000;
my $data_len                = 200;
my $run_time                = 0;
my $receive                 = 0;
my $dump                    = 0;

my %bench                   = ();
my $help                    = 0;
my $work_exit               = 0;
my $coll;
my $coll_name               = $$.'';
my $visitors                = 0;
my $rate                    = 0;

my $final_message           = 'The End!';

#-- setting up facilities

my $ret = GetOptions(
    'host=s'                    => \$host,
    'port=i'                    => \$port,
    'coll_name=s'               => \$coll_name,
    'max_size=i'                => \$max_size,
    'min_cleanup_bytes=i'       => \$min_cleanup_bytes,
    'min_cleanup_items=i'       => \$min_cleanup_items,
    'data_len=i'                => \$data_len,
    'visitors=i'                => \$max_lists,
    'run_time=i'                => \$run_time,
    'rate=f'                    => \$rate,
    'receive'                   => \$receive,
    'dump=i'                    => \$dump,
    "help|?"                    => \$help,
    );

if ( !$ret || $help )
{
    print <<'HELP';
Usage: $0 [--host="..."] [--port=...] [--coll_name="..."] [--max_size=...] [--min_cleanup_bytes=...] [--min_cleanup_items=...] [--data_len=...] [--visitors=...] [--run_time=...] [--rate=...] [--dump=...] [--receive] [--help]

Start a Redis client, connect to the Redis server, randomly inserts or receives data

Options:
    --help
        Display this help and exit.

    --host="..."
        The server should contain an IP address of Redis server.
        If not provided, '127.0.0.1' is used as the default.
    --port=N
        The server port.
        Default 6379 (the default for Redis server).
    --coll_name="..."
        Collection name.
        Default = process pid.
        To be used along with '--receive' option.
    --max_size=N
        The maximum size (bytes) of capped collection data volume.
        Default 512MB.
    --min_cleanup_bytes=N
        Size (bytes) of the data to be freed should size of the collection exceed 'max_size'.
        Default 0 - no advance cleanup performed.
    --min_cleanup_items=N
        Number of elements to be removed from collection during advance cleanup.
        Default 0 - no advance cleanup performed.
    --data_len=...
        Data unit size.
        Default 200 bytes.
    --visitors=N
        The number of lists to be created.
        Default 2_000_000.
    --run_time=...
        Maximum run time in seconds
        Default 0 - unlimited.
    --rate=N
        Exponentially distribute data using provided RATE.
        Default 0 - no exponental distribution.
    --dump=N
        Data dump volume.
        Default 0 - do not perform dump.
        WARNING: removes all keys from the current database.
    --receive
        Receive data instead of writing.
HELP
    exit 1;
}

my $item = '*' x $data_len;
my $upper = 0;
$upper = exponential( 0.99999 ) if $rate;
my $no_line = 0;

$| = 1;

#-- definition of the functions

$SIG{INT} = \&tsk;

sub tsk {
    $SIG{INT} = "IGNORE";
    $final_message = 'Execution is interrupted!';
    ++$work_exit;
    return;
}

sub client_info {
    my $redis_info = $redis->info;
    print "WARNING: Do not forget to manually delete the test data.\n";
    print '-' x 78, "\n";
    print "CLIENT INFO:\n",
        "server                 $server",                       "\n",
        "redis_version          $redis_info->{redis_version}",  "\n",
        "arch_bits              $redis_info->{arch_bits}",      "\n",
        "VISITOR_ID_LEN         ", VISITOR_ID_LEN,              "\n",
        "PORTION_TIME           ", PORTION_TIME,                "\n",
        "INFO_TIME              ", INFO_TIME,                   "\n",
        "coll_name              $coll_name",                    "\n",
        "data_len               $data_len",                     "\n",
        "maxmemory-policy       ", ( $redis->config_get( 'maxmemory-policy' ) )[1], "\n",
        "maxmemory              ", ( $redis->config_get( 'maxmemory' ) )[1],        "\n",
        "max_size               $max_size",                     "\n",
        "min_cleanup_bytes      $min_cleanup_bytes",            "\n",
        "min_cleanup_items      $min_cleanup_items",              "\n",
        "max_lists              $max_lists",                    "\n",
        "rate                   $rate",                         "\n",
        "dump                   $dump",                         "\n",
        "run_time               $run_time",                     "\n",
        '-' x 78,                                               "\n",
        ;
}

sub redis_info {
    my $redis_info = $redis->info;
    my $rss = $redis_info->{used_memory_rss};
    my $short_rss = ( $rss > 1024 * 1024 * 1024 ) ?
          $rss / ( 1024 * 1024 * 1024 )
        : $rss / ( 1024 * 1024 );
    my $info = $coll->collection_info;

    print
        "\n",
        '-' x 78,
        "\n",
        "data lists              $info->{lists}\n",
        "data items              $info->{items}\n",
        "mem_fragmentation_ratio $redis_info->{mem_fragmentation_ratio}\n",
        "used_memory_human       $redis_info->{used_memory_human}\n",
        "used_memory_rss         ", sprintf( "%.2f", $short_rss ), ( $rss > 1024 * 1024 * 1024 ) ? 'G' : 'M', "\n",
        "used_memory_peak_human  $redis_info->{used_memory_peak_human}\n",
        "used_memory_lua         $redis_info->{used_memory_lua}\n",
        "db0                     ", $redis_info->{db0} || "", "\n",
        '-' x 78,
        "\n",
        ;
}

# Send a request to Redis
sub call_redis {
    my $method      = shift;

    my @return = $redis->$method( @_ );

    return wantarray ? @return : $return[0];
}

sub exponential
{
    my $x = shift // rand;
    return log(1 - $x) / -$rate;
}

# Outputs random value in [0..LIMIT] range exponentially
# distributed with RATE. The higher is rate, the more skewed is
# the distribution (try rates from 0.1 to 3 to see how it changes).
sub get_exponentially_id {
# exponential distribution may theoretically produce big numbers
# (with low probability though) so we have to make an upper limit
# and treat all randoms greater than limit to be equal to it.

    my $r = exponential(); # exponentially distributed floating random
    $r = $upper if $r > $upper;
    return int( $r / $upper * ( $max_lists - 1 ) ); # convert to integer in (0..limit) range
}

sub measurement_info {
    my $measurement    = shift;

    print
        sprintf( "mean %.2f op/sec, mem %7s, rate has fallen by ",
            $measurement->{mean},
            $measurement->{used_memory},
            ),
        $measurement->{fallen} eq 'N/A' ? 'N/A' : sprintf( "%.2f%%", $measurement->{fallen} ),
        "\n";
}

sub _get_value {
    my $coll    = shift;
    my $key     = shift;
    my $type    = shift;

    return $coll->_call_redis( 'GET', $key )            if $type eq 'string';
    return $coll->_call_redis( 'LRANGE', $key, 0, -1 )  if $type eq 'list';
    return $coll->_call_redis( 'SMEMBERS', $key )       if $type eq 'set';

    if ( $type eq 'zset' ) {
        my %hash;
        my @zsets = $coll->_call_redis( 'ZRANGE', $key, 0, -1, 'WITHSCORES' );
        for ( my $loop = 0; $loop < scalar( @zsets ) / 2; $loop++ )
        {
            my $value = $zsets[ $loop * 2 ];
            my $score = $zsets[ ( $loop * 2 ) + 1 ];
            $hash{ $score } = $value;
        }
        return [ {%hash} ];
    }

    if ( $type eq 'hash' ) {
        my %hash;
        foreach my $item ( $coll->_call_redis( 'HKEYS', $key ) )
        {
            $hash{ $item } = $coll->_call_redis( 'HGET', $key, $item );
        }
        return {%hash};
    }
}

sub redis_dump {
    my $coll    = shift;

    my $redis_info = $redis->info;
    my %keys;

    printf( "used_memory %dB / %s\n", $redis_info->{used_memory}, $redis_info->{used_memory_human} );

    foreach my $key ( $coll->_call_redis( 'KEYS', '*' ) ) {
        my $type = $coll->_call_redis( 'TYPE', $key );
        my $show_name   = $key;
        my $encoding    = $coll->_call_redis( 'OBJECT', 'ENCODING', $key );
        my $refcount    = $coll->_call_redis( 'OBJECT', 'REFCOUNT', $key );
        $show_name     .= " ($type/$encoding: $refcount)";

        $keys{ $show_name } = _get_value( $coll, $key, $type );
    }
    return \%keys;
}

# INSTRUCTIONS -----------------------------------------------------------------

$server = "$host:$port";
$redis  = Redis->new( server => $server );
my $current_maxmemory_policy = ( $redis->config_get( 'maxmemory-policy' ) )[1];
my $current_maxmemory = ( $redis->config_get( 'maxmemory' ) )[1];

$redis->config_set( maxmemory => $max_size );
$redis->config_set( 'maxmemory-policy' => 'noeviction' );

$coll = Redis::CappedCollection->create(
    redis   => $redis,
    name    => $coll_name,
    $receive ? () : (
        min_cleanup_bytes   => $min_cleanup_bytes,
        min_cleanup_items   => $min_cleanup_items,
    ),
);

client_info();

if ( $receive and $coll_name eq $$.'' )
{
    print "There is no data to be read\n";
    goto THE_END;
}

my @measurements = ();

$redis->flushall if $dump;

my $measurement = {
    speed       => [],
    mean        => 0,
    fallen      => 0,
    used_memory => '',
    };

my $list_id;
my ( $secs, $count ) = ( 0, 0 );
my ( $time_before, $time_after );
my ( $start_time, $last_stats_reports_time, $last_info_reports_time );

my $data_id = 0;

$start_time = $last_stats_reports_time = $last_info_reports_time = time;
while ( !$work_exit )
{
    if ( $run_time ) { last if gettimeofday - $start_time > $run_time; }
    last if $work_exit;

    my $id          = ( !$receive && $rate ) ? get_exponentially_id() : int( rand $max_lists );
    $list_id        = sprintf( '%0'.VISITOR_ID_LEN.'d', $id );
    $time_before    = gettimeofday;
    my @ret         = $receive ? $coll->receive( $list_id ) : $coll->insert( $list_id, $data_id++, $item );
    $time_after     = gettimeofday;
    $secs += $time_after - $time_before;
    ++$count;

    if ( $dump and $count >= $dump )
    {
        my $dump = redis_dump( $coll );
        $Data::Dumper::Sortkeys = 1;
        print Dumper( $dump );
        goto THE_END;
    }

    my $redis_info = $redis->info;
    $measurement->{used_memory} = $redis_info->{used_memory_human},

    my $time_from_stats = $time_after - $last_stats_reports_time;
    if ( $time_from_stats > PORTION_TIME )
    {
        my $speed = int( $count / $secs );
        print '[', scalar localtime, '] ',
            $receive ? 'reads' : 'inserts', ', ',
            $secs ? sprintf( '%d', $speed ) : 'N/A', ' op/sec ',
            ' ' x 5, "\r";

        if ( gettimeofday - $last_info_reports_time > INFO_TIME )
        {
            redis_info();
            $last_info_reports_time = gettimeofday;
        }

        $secs  = 0;
        $count = 0;
        $last_stats_reports_time = gettimeofday;
    }
}

goto THE_FINISH if $work_exit;

# POSTCONDITIONS ---------------------------------------------------------------

THE_FINISH:

print
    "\n",
    $final_message,
    "\n",
    ;
THE_END:
$redis->config_set( maxmemory => $current_maxmemory );
$redis->config_set( 'maxmemory-policy' => $current_maxmemory_policy );
$redis->quit;

exit;


Makefile.PL
use 5.010;
use strict;
use warnings;

use File::Basename;
use ExtUtils::MakeMaker qw( prompt );

my %config = (
    NAME                        => 'Redis::CappedCollection',
    VERSION_FROM                => 'lib/Redis/CappedCollection.pm',
    MIN_PERL_VERSION            => '5.010',
# more to the list $Module::CoreList::version{5.010}
    PREREQ_PM                   => {
        'Const::Fast'           => '0.014',
        'Digest::SHA1'          => '2.13',
        'Mouse'                 => 0,
        'Params::Util'          => '1.07',
        'Redis'                 => '1.976',
        'Try::Tiny'             => '0.18',
        },
    TEST_REQUIRES               => {
        'Data::UUID'            => '1.217',
        'JSON::XS'              => '3.01',
        'Net::EmptyPort'        => '0',
        'Sys::SigAction'        => '0.21',
        'Test::Exception'       => '0.31',
        'Test::NoWarnings'      => '1.04',
        'Test::RedisServer'     => '0.04',
        },
    META_MERGE          => {
        no_index                => {
            directory               => [
#                't',
#                'inc',
                'xt',
            ]
        },
        provides                => {
            'Redis::CappedCollection'   => {
                file                        => 'lib/Redis/CappedCollection.pm',
                version                     => '1.09'
            },
            'Redis::CappedCollection::Util' => {
                file                        => 'lib/Redis/CappedCollection/Util.pm',
                version                     => '1.09'
            },
        },
        'meta-spec'             => { version => 2 },
        resources               => {
            repository              => {
                type                    => 'git',
                url                     => 'git://github.com/TrackingSoft/Redis-CappedCollection.git',
                web                     => 'http://github.com/TrackingSoft/Redis-CappedCollection',
            },
        },
    },
    ( $] >= 5.005 ?
        (
        ABSTRACT        => 'Redis::CappedCollection module provides the fixed sized collections that have a auto-FIFO age-out feature.',
        AUTHOR          => 'TrackingSoft LLC <sgladkov@trackingsoft.com>'
        ) : ()
    ),
    dist        => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean       => { FILES => "Debian_CPANTS.txt Redis-*.tar.gz" },
    LICENSE     => 'perl',
#    NO_MYMETA   => 1,
);

my @patterns = qw( t/*.t t/*/*.t );
$config{test} = { TESTS => join ' ', map { glob } @patterns };

if( $ENV{AUTHOR_TESTS} ) {
    @patterns = qw( xt/author/*.t xt/author/*/*.t );
    $config{test}->{TESTS} .= ' '.join ' ', map { glob } @patterns;
}

if( $ENV{RELEASE_TESTS} ) {
    @patterns = qw( xt/release/*.t xt/release/*/*.t );
    $config{test}->{TESTS} .= ' '.join ' ', map { glob } @patterns;
}

mkdir 'inc' unless -d 'inc';

ExtUtils::MakeMaker::WriteMakefile( %config );
